presto-main/src/main/java/com/facebook/presto/server/remotetask/ContinuousBatchTaskStatusFetcher.java

```diff
@@ -0,0 +1,185 @@
+import io.airlift.units.Duration;
+
+import javax.annotation.PostConstruct;
+import javax.annotation.concurrent.GuardedBy;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Consumer;
+
+import static com.facebook.airlift.http.client.HttpUriBuilder.uriBuilderFrom;
+import static com.facebook.airlift.http.client.Request.Builder.prepareGet;
+import static com.facebook.presto.client.PrestoHeaders.PRESTO_CURRENT_STATE;
+import static com.facebook.presto.client.PrestoHeaders.PRESTO_MAX_WAIT;
+import static com.facebook.presto.server.RequestHelpers.setContentTypeHeaders;
+import static com.facebook.presto.server.smile.AdaptingJsonResponseHandler.createAdaptingJsonResponseHandler;
+import static com.facebook.presto.server.smile.FullSmileResponseHandler.createFullSmileResponseHandler;
+import static com.facebook.presto.server.smile.JsonCodecWrapper.unwrapJsonCodec;
+import static com.facebook.presto.spi.StandardErrorCode.REMOTE_TASK_MISMATCH;
+import static com.facebook.presto.util.Failures.REMOTE_TASK_MISMATCH_ERROR;
+import static io.airlift.units.Duration.nanosSince;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class ContinuousBatchTaskStatusFetcher
+{
+    public class Task {
+        TaskId taskId;
+        Consumer<Throwable> onFail;
+        StateMachine<TaskStatus> taskStatus;
+        Codec<TaskStatus> taskStatusCodec;
+        Duration refreshMaxWait;
+        Executor executor;
+        HttpClient httpClient;
+        RequestErrorTracker errorTracker;
+        RemoteTaskStats stats;
+        boolean isBinaryTransportEnabled;
+
+        AtomicLong currentRequestStartNanos = new AtomicLong();
+    }
+
+    // TaskId -> WorkerId (String(URI(TaskStatus.getSelf() (get worker id)))
+    private final ConcurrentHashMap<TaskId, String> idWorkerMap;
+    // String(URI(TaskStatus.getSelf() (get worker id))) -> WorkerTaskStatusFetcher (List<Task>)
+    private final ConcurrentHashMap<String, WorkerTaskStatusFetcher> workerTaskMap;
+
+    private static final Logger log = Logger.get(ContinuousBatchTaskStatusFetcher.class);
+
+    private final AtomicLong currentRequestStartNanos = new AtomicLong();
+
+    @GuardedBy("this")
+    private boolean running;
+
+    @GuardedBy("this")
+    private ListenableFuture<BaseResponse<TaskStatus>> future;
+
+    public ContinuousBatchTaskStatusFetcher() {
+        idWorkerMap = new ConcurrentHashMap<>();
+        workerTaskMap = new ConcurrentHashMap<>();
+    }
+
+    @PostConstruct
+    public void addTask(
+            Consumer<Throwable> onFail,
+            TaskId taskId,
+            TaskStatus initialTaskStatus,
+            Duration refreshMaxWait,
```

Comment:
i don't think we need refreshMaxWait for each task as it will be per worker given we will be doing it a batch call. So may be have it as a class level attribute and passed in the constructor instead.

Reply:
Moved to constructor